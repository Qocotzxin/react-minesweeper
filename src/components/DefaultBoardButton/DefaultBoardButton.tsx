import { Box, CircularProgress } from "@mui/material";
import { FC, KeyboardEvent, SyntheticEvent, useState } from "react";
import { HiFlag } from "react-icons/hi";
import { Coordinates } from "../../types";

interface DefaultBoardButtonProps {
  colIndex: number;
  rowIndex: number;
  disabled?: boolean;
  onClick: (coordinates: Coordinates) => void;
}

export const DefaultBoardButton: FC<DefaultBoardButtonProps> = ({
  colIndex,
  rowIndex,
  disabled,
  onClick,
}) => {
  const [isLoading, setIsLoading] = useState(false);
  const [hasFlag, setHasFlag] = useState(false);

  /**
   * Executes the click prop callback passing the coordinates.
   */
  const onTileClick = () => {
    setHasFlag(false);
    setIsLoading(true);
    onClick({ x: colIndex, y: rowIndex });
  };

  /**
   * Adds a flag to the block when right clicking.
   */
  const onContextMenu = (event: SyntheticEvent) => {
    event.preventDefault();
    setHasFlag((flagged) => !flagged);
  };

  const onChange = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter") {
      onTileClick();
    }
  };

  return (
    <Box
      aria-label={`Tile ${colIndex} ${rowIndex}`}
      onKeyUp={onChange}
      tabIndex={0}
      data-testid="DefaultBoardButton"
      bgcolor="rgb(148, 151, 156)"
      border="solid #808080"
      width="1.5rem"
      height="1.5rem"
      display="flex"
      justifyContent="center"
      alignItems="center"
      onClick={onTileClick}
      onContextMenu={onContextMenu}
      sx={{
        position: "relative",
        borderWidth: " 0 1px 1px 0",
        boxShadow: 5,
        cursor: "pointer",
        pointerEvents: disabled ? "none" : "initial",
        "&:hover": {
          background: "#d0d0d0",
        },
      }}
    >
      {hasFlag && (
        <HiFlag
          size={18}
          color="rgb(201, 91, 118)"
          data-testid="DefaultBoardButton-flag"
        />
      )}
      {isLoading && <CircularProgress color="secondary" size={18} />}
    </Box>
  );
};
