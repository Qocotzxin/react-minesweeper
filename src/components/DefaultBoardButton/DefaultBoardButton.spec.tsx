import { fireEvent, render, screen } from "@testing-library/react";
import { DefaultBoardButton } from "./DefaultBoardButton";

const mockOnClick = jest.fn();

const defaultBoardButtonMockProps = {
  colIndex: 1,
  rowIndex: 2,
  onClick: mockOnClick,
};

afterEach(() => {
  mockOnClick.mockClear();
});

afterAll(() => {
  jest.resetAllMocks();
});

describe("DefaultBoardButton", () => {
  describe("Snapshot", () => {
    it("Should render correctly and match snapshot.", () => {
      const { container } = render(
        <DefaultBoardButton {...defaultBoardButtonMockProps} />
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("Behavior", () => {
    it("Should call onClick with the right coordinates.", () => {
      render(<DefaultBoardButton {...defaultBoardButtonMockProps} />);

      fireEvent.click(screen.getByTestId("DefaultBoardButton"));

      expect(mockOnClick).toHaveBeenCalledTimes(1);
      expect(mockOnClick).toHaveBeenCalledWith({ x: 1, y: 2 });
    });

    it("Should display a spinner when the button is clicked.", () => {
      render(<DefaultBoardButton {...defaultBoardButtonMockProps} />);

      fireEvent.click(screen.getByTestId("DefaultBoardButton"));

      expect(screen.getByRole("progressbar")).toBeInTheDocument();
    });

    it("Should set pointerEvents to none when the button is disabled.", () => {
      render(<DefaultBoardButton {...defaultBoardButtonMockProps} disabled />);

      expect(screen.getByTestId("DefaultBoardButton")).toHaveStyle({
        pointerEvents: "none",
      });
    });

    it("Should add a flag if available when right clicking on element.", () => {
      render(<DefaultBoardButton {...defaultBoardButtonMockProps} disabled />);

      fireEvent.contextMenu(screen.getByTestId("DefaultBoardButton"));

      expect(screen.getByTestId("DefaultBoardButton-flag")).toBeInTheDocument();
    });
  });
});
