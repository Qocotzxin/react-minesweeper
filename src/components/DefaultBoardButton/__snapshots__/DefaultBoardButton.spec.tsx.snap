// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultBoardButton Snapshot Should render correctly and match snapshot. 1`] = `
<div>
  <div
    aria-label="Tile 1 2"
    class="MuiBox-root css-7wl9ds"
    data-testid="DefaultBoardButton"
    tabindex="0"
  />
</div>
`;
